drivers: rtc: rpi_pico: Add support for the Raspberry Pi Pico RTC

ajf58 opened the pull request:
This adds the minimal get_time/set_time support for the RP2040 SOC  and enables support by default on the Pico boards. This doesn't support configuring the clock source or alarm interrupts yet.

ajf58 committed:
drivers: rtc: rpi_pico: Add support for the Raspberry Pi Pico RTC

This adds the minimal get_time/set_time support for the rp2040 and
enables support by default on the Pico boards. This doesn't support
configuring the clock source or alarm interrupts yet.

Signed-off-by: Andrew Featherstone <andrew.featherstone@gmail.com>

bjarki-trackunit commented on drivers/rtc/rtc_rpi_pico.c:
```diff
@@ -0,0 +1,102 @@
+{
+	struct rtc_rpi_pico_data *data = dev->data;
+
+	k_mutex_init(&data->lock);
```
You should use a k_spinlock instead of a mutex since you don't need to relock or track ownership of the lock :)

bjarki-trackunit commented on drivers/rtc/rtc_rpi_pico.c:
```diff
@@ -0,0 +1,102 @@
+#define TM_YEAR_REF 1900
+
+struct rtc_rpi_pico_data {
+	struct k_mutex lock;
```
Use spinlock :)

bjarki-trackunit commented on drivers/rtc/rtc_rpi_pico.c:
```diff
@@ -0,0 +1,102 @@
+{
+	datetime_t time;
+	if (!rtc_get_datetime(&time)) {
+		return -EINVAL;
```
Return -ENODATA if there is no valid time to get :)

bjarki-andreasen reviewed (changes requested):
Overall it looks good, I have requested minor changes, and will request you run the [RTC test suite](https://docs.zephyrproject.org/latest/hardware/peripherals/rtc.html#rtc-device-driver-test-suite) if you haven't already to validate the driver :) (and fix the compliance issues naturally)

bjarki-trackunit commented on drivers/rtc/rtc_rpi_pico.c:
```diff
@@ -0,0 +1,97 @@
+
+static int rtc_rpi_pico_get_time(const struct device *dev, struct rtc_time *timeptr)
+{
+	datetime_t time;
```
rename this from time to dt to avoid "reserved name" violation :)

bjarki-trackunit commented on drivers/rtc/rtc_rpi_pico.c:
```diff
@@ -0,0 +1,97 @@
+		.sec = timeptr->tm_sec,
+	};
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
```
add blank line here

bjarki-trackunit commented on drivers/rtc/rtc_rpi_pico.c:
```diff
@@ -0,0 +1,97 @@
+static int rtc_rpi_pico_get_time(const struct device *dev, struct rtc_time *timeptr)
+{
+	datetime_t time;
+	if (!rtc_get_datetime(&time)) {
```
Add blank line here

yonsch commented:
This should be merged only after #62186 has been merged, so that it uses the new clock definitions

bjarki-trackunit commented:
> This should be merged only after #62186 has been merged, so that it uses the new clock definitions

Added the DNM label :)

ajf58 committed:
drivers: rtc: rpi_pico: Add alarm support to RPi Pico RTC driver

This adds support for the alarm functionality of the RPi Pico RTC.

Signed-off-by: Andrew Featherstone <andrew.featherstone@gmail.com>

raveious commented:
Looks like #62186 has merged, can we keep this going?

ajf58 commented:
> Looks like #62186 has merged, can we keep this going?

Now that I've finished work for the year, yes. Locally I've rebased onto the `HEAD` of `main` to bring in the new clock controller work, and I think I've made the relevant other changes to my driver. I'll test those changes locally before pushing anything to my branch.

yonsch commented on drivers/rtc/rtc_rpi_pico.c:
```diff
@@ -0,0 +1,298 @@
+static int rtc_rpi_pico_alarm_get_time(const struct device *dev, uint16_t id, uint16_t *mask,
+				       struct rtc_time *timeptr);
+struct rtc_rpi_pico_data {
+	struct k_spinlock lock;
```
Any particular reason for using a spinlock and not a mutex? I see the stm32 driver uses a mutex, so it might be possible here too.

ajf58 replied:
The feedback I received previously at https://github.com/zephyrproject-rtos/zephyr/pull/64939#discussion_r1386162310 was to use a spinlock and not mutex. Can you add more detail on why you think this should be the other way around?

yonsch commented on drivers/rtc/rtc_rpi_pico.c:
```diff
@@ -0,0 +1,298 @@
+}
+#endif
+
+static int rtc_rpi_pico_init(const struct device *dev)
```
Shouldn't the clock for the RTC be enabled here somewhere?

ajf58 replied:
I created this PR before the clock controller work had been completed and merged onto `main`. Now that this is in place, I've rebased my work on the current `HEAD` of `main` and believe I have called the relevant parts of the `clock_control_` API.

ajf58 commented:
I think this is ready for another round of review feedback. This driver now includes both the time and the alarm API.

## Testing
I've run this with the RTC API test suite, and got these results:
```
*** Booting Zephyr OS build zephyr-v3.5.0-3468-g897f8e5b481e ***
Running TESTSUITE rtc_api
===================================================================
START - test_alarm
I: Setting alarm
I: Setting alarm
I: Setting alarm
 PASS - test_alarm in 26.006 seconds
===================================================================
START - test_alarm_callback
I: Setting alarm
I: Setting alarm
 PASS - test_alarm_callback in 26.004 seconds
===================================================================
START - test_set_get_time
 PASS - test_set_get_time in 0.001 seconds
===================================================================
START - test_time_counting
 PASS - test_time_counting in 9.003 seconds
===================================================================
START - test_y2k

    Assertion failed at WEST_TOPDIR/zpico/tests/drivers/rtc/rtc_api/src/test_y2k.c:59: rtc_api_test_y2k: (rtm[Y2K].tm_sec not equal to SECONDS_AFTER)
wrong second: 2
 FAIL - test_y2k in 2.015 seconds
===================================================================
TESTSUITE rtc_api failed.

------ TESTSUITE SUMMARY START ------

SUITE FAIL -  80.00% [rtc_api]: pass = 4, fail = 1, skip = 0, total = 5 duration = 63.029 seconds
 - PASS - [rtc_api.test_alarm] duration = 26.006 seconds
 - PASS - [rtc_api.test_alarm_callback] duration = 26.004 seconds
 - PASS - [rtc_api.test_set_get_time] duration = 0.001 seconds
 - PASS - [rtc_api.test_time_counting] duration = 9.003 seconds
 - FAIL - [rtc_api.test_y2k] duration = 2.015 seconds

------ TESTSUITE SUMMARY END ------

===================================================================
PROJECT EXECUTION FAILED
```
the `test_y2k` failure is something I'm not sure about; I'm wondering if this test needs the same use of `RTC_TEST_GET_SET_TIME_TOL` as found elsewhere and I'd welcome some thoughts.

bjarki-trackunit commented:
Regarding the Y2K test, it looks like the tolerance is too low for the pico, (it is taking about a second to get/set the time). Nothing to worry about :)

yonsch reviewed (dismissed):
LGTM, thanks!

bjarki-trackunit commented on drivers/rtc/rtc_rpi_pico.c:
```diff
@@ -0,0 +1,315 @@
+	}
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	// Only add to setup if it isn't -1
```
You must use /**/ style comments

bjarki-trackunit commented on drivers/rtc/rtc_rpi_pico.c:
```diff
@@ -0,0 +1,315 @@
+				       ? (((uint32_t)alarm->tm_sec) << RTC_IRQ_SETUP_1_SEC_LSB)
+				       : 0);
+
+	// Set the match enable bits for things we care about
```
/**/

bjarki-trackunit commented on drivers/rtc/rtc_rpi_pico.c:
```diff
@@ -0,0 +1,315 @@
+	data->alarm_mask = mask;
+	k_spin_unlock(&data->lock, key);
+
+	// Enable the IRQ at the peri
```
/**/

bjarki-trackunit commented on drivers/rtc/Kconfig:
```diff
@@ -51,5 +51,6 @@ source "drivers/rtc/Kconfig.mc146818"
 source "drivers/rtc/Kconfig.sam"
 source "drivers/rtc/Kconfig.stm32"
 source "drivers/rtc/Kconfig.smartbond"
+source "drivers/rtc/Kconfig.rpi_pico"
```
this should be in alphabetical order

ajf58 replied:
OK. Noting that lines 52 and 53 aren't in order, would you like me to correct that too? There's a bit of an inconsistency in the codebase between "add things chronologically" vs "add things alphabetically".

bjarki-andreasen reviewed (changes requested):
Small nits then we are good to go :)

ajf58 commented on drivers/rtc/rtc_rpi_pico.c:
```diff
@@ -0,0 +1,316 @@
+/*
+ * Copyright (c) 2020 Raspberry Pi (Trading) Ltd.
```
Note to reviewers: I've added this copyright notice here and I'd be grateful if someone to check they're happy with this. https://github.com/raspberrypi/pico-sdk/blob/master/src/rp2_common/hardware_rtc/rtc.c#L147 is in part a source of some of the work here. I think this is ok from a licensing perspective, and given there's only a very small number of ways of setting bits in register.

ajf58 commented:
Hi there,

Is there any more work I need to do to get this merged? Thanks in advance :-).

bjarki-trackunit commented:
The only blocker is the copyright clause, and I do have a way to solve it :)

Including the copyright notice in the `rtc_rpi_pico.c` is a no-go, no one even attempted to weigh in on the issue when brought up on Discord, so its out :)

However, the fix is quite simple, don't copy the code ;)

Simply refactor your driver to mirror this driver this in-tree driver https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/rtc/rtc_sam.c , take specific note of https://github.com/zephyrproject-rtos/zephyr/blob/5dad8d72ee739df5ef1b0369a5cf45ff800dec12/drivers/rtc/rtc_sam.c#L61-L94

ajf58 committed:
drivers: rtc: Create utility function for time validation

RTC drivers should validate the `struct rtc_time`'s contents against the
provided `mask`. Promote this common code to a new rtc_utils file and
modify existing drivers to use this functionality. Extend the test
coverage to include verifying this behaviour.

This is groundwork ahead of adding support for the RP2040's (as used in
the Raspberry Pi Pico) RTC and alarm.

Signed-off-by: Andrew Featherstone <andrew.featherstone@gmail.com>

ajf58 commented:
Thanks for the feedback, @bjarki-trackunit . I've refactored out the validation into a new `rtc_utils.c`, so that this can be shared across the RTC drivers rather than duplicate the code. I've extended the test cases to verify this, and also refactored some of my new code. The test cases continue to work for me. I've not tested the `rtc_sam` or `rtc_emul` changes locally, but the CI pipeline logs good.

bjarki-trackunit commented:
Nice, the `rtc_utils` are great :) I would like an extra test suite in `tests/drivers/rtc/rtc_utils` to validate the `rtc_utils_validate_tm` function on its own as well :) and I think the function `rtc_utils_validate_tm` should be named `rtc_utils_validate_rtc_time` since tm could refer to the `struct tm` which does allow -1 values :)

ajf58 commented:
> I would like an extra test suite in `tests/drivers/rtc/rtc_utils` to validate the `rtc_utils_validate_tm` function on its own as well :) and I think the function `rtc_utils_validate_tm` should be named `rtc_utils_validate_rtc_time` since tm could refer to the `struct tm` which does allow -1 values :)

I've done this as requested.

soburi reviewed (changes requested):
Changes are LGTM.
Please squash and cleanup commits.

ajf58 commented:
I did a quick correction and rebase to address a code-style issue. The CI pipeline failures are due to the runners disappearing:
> The self-hosted runner: zephyr-runner-linux-x64-4xlarge-deployment-g2tcl-r7ttn lost communication with the server.
which I don't think is down to me adding a blank line in a source file ;-).

How do we proceed? Does this require someone with higher privileges to re-run the pipeline, for example?

soburi commented on drivers/rtc/rtc_rpi_pico.c:
```diff
@@ -0,0 +1,321 @@
+	}
+
+	K_SPINLOCK(&data->lock) {
+		ret = data->alarm_pending ? 1 : 0;
```
There should be no problem with direct assignment.

soburi commented on drivers/rtc/rtc_rpi_pico.c:
```diff
@@ -0,0 +1,321 @@
+	struct rtc_rpi_pico_data *data = dev->data;
+	int err = 0;
+
+	if (timeptr->tm_year + TM_YEAR_REF > 4095) {
```
Please define this number

soburi commented on drivers/rtc/rtc_rpi_pico.c:
```diff
@@ -0,0 +1,321 @@
+	struct rtc_rpi_pico_data *data = dev->data;
+	int err = 0;
+
+	if (timeptr->tm_year + TM_YEAR_REF > 4095) {
```
Please define this number.

ajf58 replied:
Done.

bjarki-trackunit replied:
@soburi It is expected for authors to resolve conversations if the change matches the request exactly :) I do get your point though :)

ajf58 commented:
Can we get this over the line please?
